### PR TITLE
Create restapi endpoint for counting full_types

### LINKED ISSUE #4247: Create endpoint for statistics that contains process_type
To correctly calculate the statistics for Materials Cloud [1], @giovannipizzi suggests to create an endpoint:
/process_statistics or /full_types_statistics 
similar to /full_types [2] and that should also display the counts.

[1] https://www.materialscloud.org/home#statistics
[2] https://dev-aiida-dev.materialscloud.org/autowannier/api/v4/nodes/full_types

### LINKED ISSUE #4541: REST API full_type filter not translated correctly to database query filters
- [x] [AiiDA Troubleshooting Documentation](https://aiida.readthedocs.io/projects/aiida-core/en/latest/intro/troubleshooting.html)
- [x] [AiiDA Users Forum](https://groups.google.com/forum/#!forum/aiidausers)

### Describe the bug

If I do a request to the REST API ``/nodes`` endpoint, filtering by ``full_type``, the REST API doesn't translate the request to the correct query on the database whenever I have process types with unregistered entry points and the request will return an empty list.

### Steps to reproduce

1. I've tested this with the databases on the Materials Cloud. For instance, you can do a request on the curated-cofs database to get all nodes with a ``full_type`` equal to ``process.workflow.workchain.WorkChainNode.|aiida_cp2k.workchains.base.Cp2kBaseWorkChain``(click [here](https://dev-aiida-dev.materialscloud.org/curated-cofs/api/v4/nodes?full_type=%22process.workflow.workchain.WorkChainNode.|aiida_cp2k.workchains.base.Cp2kBaseWorkChain%22))
2. If I run the corresponding query on the database using the QueryBuilder:
`QueryBuilder().append(Node, filters={'node_type': {'==': 'process.workflow.workchain.WorkChainNode.'}, 'process_type': {'==': 'aiida_cp2k.workchains.base.Cp2kBaseWorkChain'}}, project=['uuid']).all()`, 
then I get a list of 1192 nodes.                                                                                                                                                                                       

### Expected behavior

The REST API call should return the list of 1192 nodes returned by the database query instead of an empty list.

### Your environment

- Operating system [e.g. Linux]: Ubuntu 18.04
- Python version [e.g. 3.7.1]: Python 3.6.9
- aiida-core version [e.g. 1.2.1]: 1.4.2

Other relevant software versions, e.g. Postres & RabbitMQ

### Additional context

The ``full_type`` filter works fine for registered entry points, e.g. [here](https://dev-aiida-dev.materialscloud.org/curated-cofs/api/v4/nodes?full_type=%22process.calculation.calcjob.CalcJobNode.|aiida.calculations:cp2k%22).

This bug affects the Materials Cloud Explore section: for instance [here](https://dev-www.materialscloud.org/explore/curated-cofs/grid/calculations) (development server) if on the left menu you click on the arrows to expand `Workflow > Work Chain > Unregistered > Aiida_cp2k > Workchains > Base > Cp2kBaseWorkChain`, and then you click on `Cp2kBaseWorkChain`, the filtered results will be empty.

### PR DESCRIPTION
**Pull request by ramirezfranciscof**
Fixes #4247 
Fixes #4541

Work in progress: currently only the end leaves are counted, and not the container branches.

- Changes in `aiida/restapi/api.py` and `aiida/restapi/common/utils.py` only add the new endpoint to relevant lists, and `aiida/restapi/resources.py` redirects the request to the new endpoint to an internal function.
- The internal function (`get_namespace`) resides in `aiida/restapi/translator/nodes/node.py`, and is the same as the one requested for the `full_types` endpoint, except it has been adapted with an optional keyword to count the nodes. This in turn is passed to `get_node_namespace` inside of `aiida/restapi/common/identifiers.py`.
- Most modifications where done in `aiida/restapi/common/identifiers.py`, where the namespace class is defined. The initialization was adapted to include a counter, which receives an `None` value unless the option to count nodes was selected when calling `get_node_namespace`. This value gets returned when the `get_description` method is called (when requesting the endpoint).

### CODE REVIEW

**Comment by ramirezfranciscof** [comment]
For reference: I just tried this with the 2D materials database and the results seem to indicate that the time of the `full_types` entrypoint shouldn't have changed much, whereas the time of the `full_types_count` should be at most ten times slower. Here are a few plots of 100 calls to the `get_namespace` method (for the `develop` branch, this PR with `count=False` and this PR with `count=True` respectively):

results_develop
`develop` branch

results_false
PR with `count=False`

results_true
PR with `count=True`

The numbers remained mostly the same for django or sqlalchemy backends. I'll see if I can try some bigger database.

**Comment by ramirezfranciscof** [comment]
Ok, I was checking on the status of this and realized the work in progress (WIP) tag set by @CasperWA . I'm not sure exactly what would be interpreted as a WIP but just in case let me clarify the current situation:

1. **Re testing it with the materials cloud:** this would be the next step, I already asked @vgranata if she could do it. Also, for valeria: in case you saw the WIP tag and thought I was still working on this, maybe also because of my last comment (see 3), this is not the case, you have green light to do the test.
2. **Re code reviewing:** I ideally wanted to check 1 before doing a proper code review. If that is the meaning of the WIP (the website check being the "work"), then I guess its ok to keep it.
3. **Re my last comment:** in the end I wasn't still able to get the bigger database (I am getting a "not enough space" error when trying to import it) so that will likely not happen. Anyways I would thing that the timings with the middle sized one are enough to go on. So this should not be the reason for the WIP tag.

**Comment by vgranata** [comment]
@ramirezfranciscof Thank you Francisco for clarifying this. 
@asle85 Elsa offered to look and test this PR.

**Comment by asle85** [comment]
I've tested the new endpoint on the AiiDA databases used in Materials Cloud: it's very fast for small databases, and for the biggest ones (until ~1.4 million nodes) it takes around 3 seconds.

@giovannipizzi which level of specification do we want to show in the statistics pie chart? Is it ok to have the namespaces (or labels) of the leaves, i.e. namespaces that do not have subspaces? Does it make sense to include this information in the `nodes/statistics` endpoint?

**Comment by giovannipizzi** [comment]
Thanks! I think it's fast enough. @asle85 - I don't know, I think we need to check together a couple of examples to decide. If this PR anyway gives enough info to create any piechart, I suggest to move on and merge it, and we can discuss what to show in the pie chart in an issue on the issue tracker of one of the Materials Cloud repositories.

**Comment by ramirezfranciscof** [comment]
Ok, good; I just need now someone to review the code and I'll merge. I requested to @sphuber , but he may be busy with other stuff, so feel free to request to another person or even do it yourself if you have the time @giovannipizzi .

**Comment by ramirezfranciscof** [comment]
@flavianojs I cannot add you as reviewer not sure why (maybe you need to be part of the aiidateam and you are not?)

**Comment by chrisjsewell** [comment]
I'll look through this tomorrow @ramirezfranciscof, pester me if you don't hear anything lol

**Comment by ramirezfranciscof** [comment]
Hey @chrisjsewell : friendly reminder if you can check this.

**Comment by chrisjsewell** [comment]
Yep it generally looks fine to me.
However, given that this is a PR on counting, I think the test should actually make sure that the count is correct 😉 

"Ideally", I'd say re-write the test file as pytest (using the fixtures), then just use [pytest-regressions](https://pypi.org/project/pytest-regressions/) to validate the whole JSON blob, since its all deterministic.

But, at the very least, check that the top-level counter is 9

```json
    data = {
        'counter':
        9,
        'full_type':
        'node.%|',
        'label':
        'node',
        'namespace':
        'node',
        'path':
        'node',
        'subspaces': [{
            'counter':
            6,
            'full_type':
            'data.%|',
            'label':
            'Data',
            'namespace':
            'data',
            'path':
            'node.data',
            'subspaces': [{
                'counter':
                1,
                'full_type':
                'data.array.%|',
                'label':
                'array',
                'namespace':
                'array',
                'path':
                'node.data.array',
                'subspaces': [{
                    'counter': 1,
                    'full_type': 'data.array.kpoints.KpointsData.|',
                    'label': 'KpointsData',
                    'namespace': 'kpoints',
                    'path': 'node.data.array.kpoints',
                    'subspaces': []
                }]
            }, {
                'counter': 1,
                'full_type': 'data.cif.CifData.|',
                'label': 'CifData',
                'namespace': 'cif',
                'path': 'node.data.cif',
                'subspaces': []
            }, {
                'counter': 2,
                'full_type': 'data.dict.Dict.|',
                'label': 'Dict',
                'namespace': 'dict',
                'path': 'node.data.dict',
                'subspaces': []
            }, {
                'counter': 1,
                'full_type': 'data.folder.FolderData.|',
                'label': 'FolderData',
                'namespace': 'folder',
                'path': 'node.data.folder',
                'subspaces': []
            }, {
                'counter': 1,
                'full_type': 'data.structure.StructureData.|',
                'label': 'StructureData',
                'namespace': 'structure',
                'path': 'node.data.structure',
                'subspaces': []
            }]
        }, {
            'counter':
            3,
            'full_type':
            'process.%|%',
            'label':
            'Process',
            'namespace':
            'process',
            'path':
            'node.process',
            'subspaces': [{
                'counter':
                3,
                'full_type':
                'process.calculation.%|%',
                'label':
                'Calculation',
                'namespace':
                'calculation',
                'path':
                'node.process.calculation',
                'subspaces': [{
                    'counter': 1,
                    'full_type': 'process.calculation.calcfunction.CalcFunctionNode.|',
                    'label': 'CalcFunctionNode',
                    'namespace': 'calcfunction',
                    'path': 'node.process.calculation.calcfunction',
                    'subspaces': []
                }, {
                    'counter': 2,
                    'full_type': 'process.calculation.calcjob.CalcJobNode.|',
                    'label': 'CalcJobNode',
                    'namespace': 'calcjob',
                    'path': 'node.process.calculation.calcjob',
                    'subspaces': []
                }]
            }]
        }]
    }
```

File: aiida/restapi/common/identifiers.py
Comment (chrisjsewell) [inline aiida/restapi/common/identifiers.py:387]: there should also really be a test for this scenario

**Review by chrisjsewell — DISMISSED** [review]
See comment

**Comment by asle85** [comment]
@ramirezfranciscof Since we are adding a new endpoint, it makes sense to release a new minor version of the REST API, i.e change the version to 4.1.0 in the [config](https://github.com/aiidateam/aiida-core/blob/06063e7c38b6e40d3d2fbafe84e4e7f6adc8782a/aiida/restapi/common/config.py#L19). In this way we can detect from the version number if the endpoint will be there or not.

File: aiida/restapi/common/identifiers.py
Comment (asle85) [inline aiida/restapi/common/identifiers.py:118]: I think you forgot to remove this

**Review by ramirezfranciscof — COMMENTED** [review]
I'll leave an explanation of the most recent changes as comments in the code. The new behavior of the `full_types`/`full_types_count` functionality will

- Group together all process nodes (or any kind of node really) with `process_type = ''` and `process_type = None` into a leaf with full type like `process.calculation.calc…tion.CalcFunctionNode.|` (node type depending on the case).

- Consider all unregistered node under a branch labeled "unregistered" with no full path (so you can't query for all unregistered nodes). Each will have a full type of the form: `process.calculation.calc…ctionNode.|aiida_cp2k.%"` (note the process part of th full type does not start with `aiida.calculations:...`.

I will now prepare the test and deal with any other remaining comment from the previous review, but I just want to confirm that this behavior is ok with @giovannipizzi and @sphuber .

File: aiida/restapi/common/identifiers.py
Comment (ramirezfranciscof) [inline aiida/restapi/common/identifiers.py:106]: Moreover, I decided to remove the escaping when we were not using like because it would seem this is only for the like sql search?

```
def escape_for_sql_like(string):
    """Function that escapes % or _ symbols provided by user

    SQL LIKE syntax summary:

    - ``%`` -> match any number of characters
    - ``_`` -> match exactly one character

    """
    return string.replace('%', '\\%').replace('_', '\\_')
```

Moreover in testing this seems to be the case:

```
In [X]: qb = QueryBuilder()
In [X]: qb.append( orm.ProcessNode, filters={
In [X]:     'process_type': {'like': 'aiida\\_lsmo.workchains.isotherm.get\\_atomic\\_radii'}
In [X]:      })
In [X]: qb.count()
Out[X]: 348

In [X]: qb = QueryBuilder()
In [X]: qb.append( orm.ProcessNode, filters={
In [X]:     'process_type': 'aiida\\_lsmo.workchains.isotherm.get\\_atomic\\_radii'
In [X]:      })
In [X]: qb.count()
Out[X]: 0

In [X]: qb = QueryBuilder()
In [X]: qb.append( orm.ProcessNode, filters={
In [X]:     'process_type': 'aiida_lsmo.workchains.isotherm.get_atomic_radii'
In [X]:      })
In [X]: qb.count()
Out[X]: 348
```

Were we escaping this for any particular reason or was this a bug? @giovannipizzi @sphuber

**Comment by sphuber** [comment]
Thanks for the rebase @ramirezfranciscof that helps a lot. I will start going through it now. First question that I have now straight away: can you give me examples of the process nodes that have an empty process type string and those that have `None` (apparently)? I am not even sure how that one happens. Is there then `null` in the database column?

**Comment by ramirezfranciscof** [comment]
There should be calcfunctions with `process_type=None` in this database https://dev-aiida-dev.materialscloud.org/curated-cofs/api/v4/ and`process_type=''` in this other one https://dev-aiida-dev.materialscloud.org/2dtopo/api/v4/. The second database can be downloaded from here https://archive.materialscloud.org/record/2020.86, I don't know to which archive it corresponds the first one though, @asle85 ?

File: tests/restapi/test_identifiers.py
Comment (sphuber) [inline tests/restapi/test_identifiers.py:179]: should we also add this case to `test_get_full_type_filters`, i.e. checking there the correct output of `get_full_type_filters` for the anomaly process types?

File: aiida/restapi/common/identifiers.py
Comment (sphuber) [inline aiida/restapi/common/identifiers.py:372]:
```suggestion
               'user_id': user_pk
```

File: aiida/restapi/common/identifiers.py
Comment (sphuber) [inline aiida/restapi/common/identifiers.py:376]:
```suggestion
    filters = {}

    if user_pk is not None:
        filters['user_id'] = user_pk

    builder = orm.QueryBuilder().append(orm.Node, filters=filters, project=['node_type', 'process_type']).distinct()
```

File: aiida/restapi/resources.py
Comment (sphuber) [inline aiida/restapi/resources.py:277]: where is `headers` used?

File: aiida/restapi/resources.py
Comment (sphuber) [inline aiida/restapi/resources.py:279]:
```suggestion
                user_pk = filters['user']['==']
```

File: aiida/restapi/resources.py
Comment (sphuber) [inline aiida/restapi/resources.py:281]:
```suggestion
                user_pk = None
```

File: tests/restapi/test_statistics.py
Comment (sphuber) [inline tests/restapi/test_statistics.py:57]:
```suggestion
```

File: tests/restapi/conftest.py
Comment (sphuber) [inline tests/restapi/conftest.py:74]: why do you need all these comments and attributes/extras you set on the calculations further down, as well as the `Log` entries? They are not involved in the actual test on the namespace tree whatsoever

File: tests/restapi/conftest.py
Comment (sphuber) [inline tests/restapi/conftest.py:111]: Same here, you are not actually testing any of the files of calcjobnodes in the `test_count_consistency` test, which is the only place you use this fixture don't you?

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @ramirezfranciscof . I have some suggestions and questions. I already reviewed the first commit in a different tab, but those don't show up in this review. I hope they won't get lost and show up...

File: aiida/restapi/common/identifiers.py
Comment (ramirezfranciscof) [reply aiida/restapi/common/identifiers.py:120]: So, I don't think it would have the same effect because currently if `process_type` is **either** `''` or `None`, it will search for **both** `''` and `None`. Your change would make `None` just search for `None` and `''` just search for `''`.

The reason I grouped these together is that this function takes the `full_type` string (that looks like `node.type.Descriptor|process.type.descriptor`) and splits it into the `node_type` and the `process_type`. It is possible to pass an empty `process_type` in this way (`node.type.Descriptor|`) but it is not possible to do so with a null process type (as there is no simple way to distinguish between `' None'` string and `None` in `node.type.Descriptor|None`).

File: aiida/restapi/common/identifiers.py
Comment (ramirezfranciscof) [reply aiida/restapi/common/identifiers.py:397]: What do you mean it doesn't do anything? This creates a whole separate branch structure for the unregistered processes. If you use just the label then `aiida_qe.base_workchain` and `aiida_siesta.base_workchain` would be considered as the same endpoint.

File: tests/restapi/test_identifiers.py
Comment (ramirezfranciscof) [reply tests/restapi/test_identifiers.py:173]: Out of curiosity, why this change? The `cls` variable gets highlighted in my editor which makes me think it could be reserved.

File: tests/restapi/test_identifiers.py
Comment (ramirezfranciscof) [reply tests/restapi/test_identifiers.py:179]: What do you mean checking the "correct output" of `get_full_type_filters`? In this test below I'm using `get_full_type_filters` to get the filters for the queries I'm using to check, thus checkign the "correctedness" in the sense that it gives filters that work on the querybuilder. Do you mean checking them literally instead of functionally? Isn't this more versatile?

File: aiida/restapi/resources.py
Comment (ramirezfranciscof) [reply aiida/restapi/resources.py:277]: I honestly have no idea, I just copied this from the previous section for `'full_types'` (and also a bit from `'statistics'`, that is why I used the `usr` variable). It is returned at the end of this function so I guess you need to set them?

File: tests/restapi/conftest.py
Comment (ramirezfranciscof) [reply tests/restapi/conftest.py:111]: No, you are right. I think originally I thought of using this fixture for all restapi tests; I copied this from the setup in `test_routes.py`, where I put the first tests of this, but then I made the separate `test_statistics.py` for which I only need the nodes. So I can't take all the rest out.

File: aiida/restapi/common/identifiers.py
Comment (sphuber) [reply aiida/restapi/common/identifiers.py:120]: so you are saying that the full types will look like `node.type|None` both when `process_type` is `''` as well as `None`? In that case, you are right that what I proposed is not the same. I am still worried about all these edge cases that really shouldn't exist and we should figure out how and why they came about. At the very least, you should add a comment here to explain that your additional clause is because there (erroneously) exist process nodes with a process_type equal to empty string or null

File: aiida/restapi/common/identifiers.py
Comment (sphuber) [reply aiida/restapi/common/identifiers.py:244]: you resolved this, but I don't see anything changed?

File: aiida/restapi/common/identifiers.py
Comment (sphuber) [reply aiida/restapi/common/identifiers.py:397]: ah, apologies, I missed that `label` was changed to `process_type`. What I meant was that `[DEFAULT_NAMESPACE_LABEL] + [process_type]` is the same as  `[DEFAULT_NAMESPACE_LABEL, process_type]` which is what it used to be and should be ever so slightly more efficient (not noticeably but still)

File: tests/restapi/test_identifiers.py
Comment (sphuber) [reply tests/restapi/test_identifiers.py:173]: just that variable names here should be preferably lower case. `cls` is not reserved but is typically used in class methods. There is nothing wrong with using it outside of that scope as long as it doesn't clash with actual class reference, which is the case here.

File: tests/restapi/test_identifiers.py
Comment (sphuber) [reply tests/restapi/test_identifiers.py:179]: in `test_get_full_type_filters` you do not add test cases for `process_type in ['', None]`, but the function contains explicit logic branches for it. Now, you *indirectly* test it through these tests here below, but that is in a roundabout way and not very direct. It should be tested as closely to the code itself as possible as to make it very clear

File: aiida/restapi/common/identifiers.py
Comment (ramirezfranciscof) [reply aiida/restapi/common/identifiers.py:120]: > so you are saying that the full types will look like `node.type|None` both when process_type is `''` as well as `None`?

Nono, it will show `node.type|` both when it is `''` and when it is `None`. I mentioned the `node.type|None` example to illustrate that it is not easy to distinguish between these two cases because you need to encode this into a string (the `full_type`) and therefore you also need to distinguish between the null `None` and the string `'None'`.

> I am still worried about all these edge cases that really shouldn't exist and we should figure out how and why they came about. At the very least, you should add a comment here to explain that your additional clause is because there (erroneously) exist process nodes with a process_type equal to empty string or null

Ok, I can add that, and we could open an issue. I mean, it might be the case that this was a bug in some migrations that already got fixed, but nevertheless these nodes will exist in databases so I think it is not bad to be able to catch these in the most securely way possible.

File: tests/restapi/test_identifiers.py
Comment (ramirezfranciscof) [reply tests/restapi/test_identifiers.py]: > just that variable names here should be preferably lower case

Ah, yes, pylint told me the same thing xD. Even if the variables really are classes? Should I then also change `ProcessClass` in the previous test?

File: tests/restapi/test_identifiers.py
Comment (ramirezfranciscof) [reply tests/restapi/test_identifiers.py]: Ok, I think I understand what you mean. I will add some more explicit tests there.

File: tests/restapi/test_identifiers.py
Comment (ramirezfranciscof) [reply tests/restapi/test_identifiers.py]: Ok, in the end I used `node_class` and `process_class`; they are lower case and to me a bit more explicit than `cls` (at least relative to one another they indicate the "limits" of what is being tested), but I can change for `cls` if you think that is better for any reason.

File: aiida/restapi/common/identifiers.py
Comment (ramirezfranciscof) [reply aiida/restapi/common/identifiers.py]: Ah, ok, haha no worries. Yeah, I think there was a middle step in which we tried to set it to something different (like removing the `DEFAULT_NAMESPACE_LABEL`), but in the end settled with this, that is why it changed unnecessarily. It is done now.

**Comment by CasperWA** [comment]
Hey, just want to say I have some corrections that should go into these files as well. I have implemented them in the #4337 PR, but I haven't had time to update it yet, so just highlighting some things that may be interesting to implement here:

https://github.com/aiidateam/aiida-core/pull/4337/files#diff-4bb11213826a021effbfd69c334387fd7c536eb5579f1f3a6bb0bd372770f671R72

Ah - and I see now I am missing uploading the stuff that touches on generating the node and process types. Since it's on my home desktop computer and I'm currently in the office, I'll have to push that tonight, if I remember. It just simplifies the creation of the `full_type` field value, ensuring it doesn't fail if any of the parts should be `None` or empty strings.

**Comment by ramirezfranciscof** [comment]
> Ah - and I see now I am missing uploading the stuff that touches on generating the node and process types. Since it's on my home desktop computer and I'm currently in the office, I'll have to push that tonight, if I remember. It just simplifies the creation of the full_type field value, ensuring it doesn't fail if any of the parts should be None or empty strings.
>
> Edit: If this information is completely out-of-scope of this PR, then please disregard it 😅

@CasperWA haha, well yes it is in scope, this is one of the changes I implemented here. Please do try to push the changes asap and ping me when you do; we were trying to get this PR merge for tomorrow's release. Do you remember what changes you made related to this that you could share now?

**Comment by CasperWA** [comment]
> > Ah - and I see now I am missing uploading the stuff that touches on generating the node and process types. Since it's on my home desktop computer and I'm currently in the office, I'll have to push that tonight, if I remember. It just simplifies the creation of the full_type field value, ensuring it doesn't fail if any of the parts should be None or empty strings.
> > Edit: If this information is completely out-of-scope of this PR, then please disregard it sweat_smile
> 
> @CasperWA haha, well yes it is in scope, this is one of the changes I implemented here. Please do try to push the changes asap and ping me when you do; we were trying to get this PR merge for tomorrow's release. Do you remember what changes you made related to this that you could share now?

I believe I simplified this function to a single f-string:
https://github.com/aiidateam/aiida-core/blob/bd197f31eb4ccf9a84ed2634cece62a74065c54a/aiida/restapi/common/identifiers.py#L64-L77

Remove the try-except here:
https://github.com/aiidateam/aiida-core/blob/bd197f31eb4ccf9a84ed2634cece62a74065c54a/aiida/restapi/translator/nodes/node.py#L566-L569
Using instead the `.get` method.

The latter change here relies on what `full_type`'s value should be if any of `node_type` or `process_type` are `None` or empty strings. If the desired result is that the value _should_ be `None`/`null`, then the `construct_full_type` function should reflect this, not the latter part, where the function is used.

Maybe you've already fixed this (didn't check, sorry)? But that's at least what I can remember for now :) I'll have to push what I have locally tonight otherwise.

**Comment by ramirezfranciscof** [comment]
@CasperWA Ok, thanks, I see you've updated your branch. I think we are both dealing with the problem of atipical `process_types` but in different parts of the restAPI code, so our changes seem to be complementary. I think my modifications shouldn't affect yours, but you can rebase once we merge this and let me know if there is any problem.

@sphuber @chrisjsewell ready for a new pass.

**Comment by CasperWA** [comment]
> @CasperWA Ok, thanks, I see you've updated your branch. I think we are both dealing with the problem of atipical `process_types` but in different parts of the restAPI code, so our changes seem to be complementary. I think my modifications shouldn't affect yours, but you can rebase once we merge this and let me know if there is any problem.

@ramirezfranciscof I was more thinking of moving some of the changes over to this PR, not if the changes in each of the PRs are compatible with each other.
The changes I have made were simply fixes to some issues I experience, i.e., they can be implemented in this PR as well if that was your wish. Especially because I saw you were touching similar aspects of the REST API code.

But I'll leave it up to you and just rebase whenever, if it comes to that :+1:

**Comment by ramirezfranciscof** [comment]
@CasperWA Yeah, we can also do that. I can check with @sphuber if he thinks its better to do that: this one has to go today, si it will depend on how much work would be to incorporate both together. For starters, we might need to rebase your branch over **the first** commit of this one (title "Further considerations for `None` or empty types"?), check that all tests work, and then I would rebase over yours. If you want maybe you can test the first rebase in a separate branch to verify that it does it automatically and all tests work and confirm here.

@chrisjsewell would you be able to give this a final pass before our meeting at 15?

**Comment by CasperWA** [comment]
> @CasperWA Yeah, we can also do that. I can check with @sphuber if he thinks its better to do that: this one has to go today, so it will depend on how much work would be to incorporate both together. For starters, we might need to rebase your branch over **the first** commit of this one (title "Further considerations for `None` or empty types"?), check that all tests work, and then I would rebase over yours. If you want maybe you can test the first rebase in a separate branch to verify that it does it automatically and all tests work and confirm here.

I think you misunderstand me. My PR should not be incorporated into this PR. But I implemented some minor fixes in my PR that may be relevant to get into this one. But it seems it will be too difficult, so let's keep it separate and I will update my PR accordingly.

File: tests/restapi/test_identifiers.py
Comment (sphuber) [reply tests/restapi/test_identifiers.py]: that's fine

**Review by sphuber — APPROVED** [review]
Thanks a lot @ramirezfranciscof

**Pull request merged**
